### PR TITLE
Add new profling options for Facebook/hhvm

### PR DESCRIPTION
**Pull request by paulbiss**
Summary: This is commit adds a hand full of convenience options for:
- Enabling/disabling the FileCache and RepoAuthoritative mode during profiling
- Generating TCDumps (optionally of the top functions/translations)
- Profiling bytecodes (as opposed to PHP functions)
- Keeping the server on online after profiling is complete
- Running repo-mode with AllVolatile enabled (hoisting is technically broken...sigh)
- Adjusting the PCRE caching scheme, and cache size
- Dumping the contents of the PCRE cache
- Interpreting all pseudomains (guard relaxation is broken for pseudomains...sigh)
- fbcode option to use an fbcode build of hhvm/tc-print (the latter being for TCDumps)

The following changes were also made:
- Run 300 serial warmup requests
- Run one round of parallel requests before benchmarking
- Write the pid file path to nginx.conf (useful for testing)
- Remove the check for connectivity in wordpress (was unnecessary, and proxies suck)
- Remove the -k option for timeout (not supported on devservers)

The following is still broken on devservers:
- Cannot create databases with MyISAM tables (not sure of the best fix but we do need to fix this)

The following features would be useful in a followup PR:
- Option to run linux perf tool
- Option to run rollup tool (integrated with fbcode option)
- Option to dump repo
- Option to dump file-cache

The additional warmup time was added after inspecting the tc-filling behavior and discovering that tc size was still dramatically increasing after 50 warmup requests. The additional serial warmup requests ensure near optimal tc layout, while the extra complete run of parallel requests is for a handful of frameworks that continue to substantially JIT even after a high volume of single threaded requests (mostly in conjunction with broken pseudomain JITing). While inconvenient the added warmup should ensure better consistency (this was certainly observed in testing). It's also particularly important in repo-auth mode where hot functions are JITed into Ahot after appropriate profiling which may further delay warmup. 

Test Plan: Tested all new options, ran hh_client

### CODE REVIEW

File: base/PerfOptions.php [hunk 1]
```diff
@@ -134,6 +180,50 @@ public function __construct(Vector<string> $argv) {
...
+      } else if ($isFacebook) {
+        $this->tcprint =
+          $fbcode . '/_bin/hphp/facebook/tools/tc-print/tc-print';
+      }
+    }
+    $this->tc_alltrans = array_key_exists('dump-all-trans', $o);
+    $this->tc_toptrans = array_key_exists('dump-top-trans', $o);
+    $this->tc_topfuncs = array_key_exists('dump-top-funcs', $o);
+    $this->pcredump    = array_key_exists('dump-pcre-cache', $o);
+    $this->profBC      = array_key_exists('profBC', $o);
```
Comment (fredemmott) [inline base/PerfOptions.php]: FB standard is:
 - locals are under_scored
 - members are camelCased

Hopefully I was consistent, but these new ones aren't even with each other.

File: base/PerfRunner.php [hunk 2]
```diff
@@ -93,6 +93,15 @@ private static function RunWithOptionsAndEngine(
     invariant(!$siege->isRunning(), 'Siege is still running :/');
     invariant($php_engine->isRunning(), get_class($php_engine).' crashed');
 
+    self::PrintProgress('Starting Siege for multi request warmup');
+    $siege = new Siege($options, $target, RequestModes::BENCHMARK);
```
Comment (fredemmott) [inline base/PerfRunner.php]: Add a new RequestMode

File: base/PerfRunner.php [hunk 3]
```diff
@@ -93,6 +93,15 @@ private static function RunWithOptionsAndEngine(
...
+    $siege->start();
+    invariant($siege->isRunning(), 'Failed to start siege');
+    $siege->wait();
```
Comment (fredemmott) [inline base/PerfRunner.php]: other than the request mode, I'm fine with this block existing, but I think an explanation is needed both here and in the readme (or here referencing the readme) saying why there's two warmup modes now.

File: base/Process.php [hunk 4]
```diff
@@ -99,19 +99,16 @@ public function isRunning(): bool {
...
     }
     if (is_resource($this->stdout)) {
       pclose($this->stdout);
+      $this->stdout = null;
     }
-    $pid = $this->getPid();
-    if ($pid !== null) {
-      posix_kill($pid, SIGTERM);
+    if (is_resource($this->process)) {
+      proc_terminate($this->process);
```
Comment (fredemmott) [inline base/Process.php]: This breaks if shell is /bin/dash, eg ubuntu machines.

File: base/Process.php [hunk 5]
```diff
@@ -128,12 +125,9 @@ public function getPid(): ?int {
   }
 
   public function wait(): void {
-    $pid = $this->getPid();
-    if ($pid === null) {
-      return;
+    if (is_resource($this->process)) {
+      proc_close($this->process);
```
Comment (fredemmott) [inline base/Process.php]: ditto.

File: base/Siege.php [hunk 6]
```diff
@@ -91,12 +91,10 @@ protected function getArguments(): Vector<string> {
     if (!$this->options->noTimeLimit) {
       $arguments = Vector {
         // See Siege::getExecutablePath()  - these arguments get passed to
-        // 'timeout'
-        '-k', '5m',
         '5m',
         parent::getExecutablePath(),
       };
-    };
+    }
```
Comment (fredemmott) [inline base/Siege.php]: Nope.

File: targets/wordpress/WordpressTarget.php [hunk 7]
```diff
@@ -78,11 +78,6 @@ public function needsUnfreeze(): bool {
   // Contact rpc.pingomatic.com, upgrade to latest .z release, other periodic
   // tasks
   public function unfreeze(PerfOptions $options): void {
-    // Need internet access or wordpress will keep on retrying this stuff
-    if (!file_get_contents('http://www.example.com')) {
-      throw new Exception(
-        'Wordpress requires internet access to http://www.example.com');
-    }
     // Does basic bookkeeping...
```
Comment (fredemmott) [inline targets/wordpress/WordpressTarget.php]: good, but should probably be a separate commit

File: base/Process.php (hunk 4 above)
Comment (paulbiss) [inline base/Process.php]: I don't think posix_kill is going to reap the subprocess. I could add a pcntl_wait after the kill, but I'm surprised proc_* isn't more portable.

File: base/Process.php (hunk 5 above)
Comment (paulbiss) [inline base/Process.php]: Was done for consistency with above

File: base/Siege.php (hunk 6 above)
Comment (paulbiss) [inline base/Siege.php]: It is GNU timeout, but it only supports -s. We could set it to just immediately send a sigkill rather than a sigterm then a sigkill with `--signal=9`

Comment (fredemmott) [inline base/Siege.php]: This isn't ideal :( we could put timeout in timeout? :p

Comment (fredemmott) [inline base/Siege.php]: from offline discussion: --signal=9 is fine, because this only happens when it's already tried to kill itself, and is stuck in a signal handling loop

File: base/HHVMDaemon.php [hunk 8]
```diff
@@ -73,11 +73,84 @@ protected function getArguments(): Vector<string> {
...
+        '-k1',
+      };
+
+      if ($this->options->allVolatile) {
+        $args->add('-v'); $args->add('AllVolatile=true');
+      }
+
+      invariant(is_dir($sourceRoot), 'Could not find valid source root');
+
+      $files = shell_exec("cd $sourceRoot && find . -type f");
```
Comment (fredemmott) [inline base/HHVMDaemon.php]: recursiveiteratoriterator + recursivedirectoryiterator would be nice

File: base/HHVMDaemon.php [hunk 9]
```diff
@@ -108,8 +182,56 @@ public function stop(): void {
...
+        file_put_contents('tc-top-trans', $toptrans);
+      }
+
+      if ($this->options->tcTopfuncs) {
+        $topfuncs = shell_exec($command . ' -T 100');
+        file_put_contents('tc-top-funcs', $topfuncs);
+      }
+    }
+
+    if ($this->options->pcredump) {
```
Comment (fredemmott) [inline base/HHVMDaemon.php]: Do we need to check /tmp/pcre_cache does not exist first?

File: base/PerfOptions.php [hunk 10]
```diff
@@ -35,7 +35,34 @@
...
+
+  //
+  // HHVM specific options to enable RepoAuthoritative mode and the static
+  // content cache, as well as selecting the PCRE caching mode.
+  //
+  public bool $precompile = false;
+  public bool $filecache = false;
+  public string $pcre_cache = "static";
+  public int $pcre_size = 98304;
+  public int $pcre_expire = 7200;
```
Comment (fredemmott) [inline base/PerfOptions.php]: still mixing camelcase and underscored members here

**Comment by fredemmott** [comment]
Given how many shell_exec() this adds, probably worth adding a helper that takes string $cmd, Vector<string> args.

File: base/Process.php (hunk 4 above)
Comment (fredemmott) [inline base/Process.php]: proc_terminate() is consistent - the problem is that what $this->process refers to isn't.

Bash just execs() in simple cases, so $this->process is nginx/hhvm/siege/whatever. Dash *always* forks and execs, so $this->process is referring to the shell. This gets us weirdness.

**Comment by fredemmott** [comment]
Can you force-push a rebase+squash, ideally add the shell_exec() function, then I'll merge this?

File: base/PerfOptions.php [hunk 11]
```diff
@@ -35,7 +35,34 @@
   public bool $dumpIsCompressed = true;
   public bool $traceSubProcess = false;
   public bool $noTimeLimit = false;
-  
+
+  // Pause once benchmarking is complete to allow for manual inspection of the
+  // HHVM or PHP process.
+  public bool $waitAtEnd = false;
```
Comment (rrh) [inline base/PerfOptions.php]: thank you for adding this!

File: base/PerfOptions.php [hunk 12]
```diff
@@ -134,6 +180,50 @@ public function __construct(Vector<string> $argv) {
...
+        $this->tcprint = $tcprint;
+      } else if ($isFacebook) {
+        $this->tcprint =
+          $fbcode . '/_bin/hphp/facebook/tools/tc-print/tc-print';
+      }
+    }
+    $this->tcAlltrans = array_key_exists('dump-all-trans', $o);
+    $this->tcToptrans = array_key_exists('dump-top-trans', $o);
+    $this->tcTopfuncs = array_key_exists('dump-top-funcs', $o);
+    $this->pcredump    = array_key_exists('dump-pcre-cache', $o);
```
Comment (rrh) [inline base/PerfOptions.php]: if you are going to use ascii art and line up the = on lines 215 and 216, also do that for lines 212 through 214, above

**Comment by fredemmott** [comment]
Going to revisit --i-am-not-benchmarking, this breaks its' purpose.
